### PR TITLE
fixes

### PR DESCRIPTION
**Pull request by Punzo**
@hackermd few fixes and additions:

1) Adds groupMonochormeMetadataInstances and groupColoredMetadataInstances functions in metadata for https://github.com/MGHComputationalPathology/slim/issues/22 and uses them in the VolumeImageViewer constructor. 
NOTE: for the grouping functions, I don't pass a key, but I just use the OpticalPathIndentifier. For passing a key to the functions, we would need to use in those functions non formatted DICOM attributes and I would rather avoid that.

2) Fix visibility when no information for channels are provided at the constructor https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/compare/master...Punzo:minorfixes?expand=1#diff-245646583d4a8774103504d6d7c723409a67b490d2adf2bf6c90fb7c6372214dR776-R783

3) fix undefined variable for error report in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/compare/master...Punzo:minorfixes?expand=1#diff-245646583d4a8774103504d6d7c723409a67b490d2adf2bf6c90fb7c6372214dR730

4) For overimposing multi-channels, we were actually not checking for the FrameOfReferenceUID attribute. I added the  FrameOfReferenceUID checks as well, but I had to comment them. This because the test data set (http://34.68.90.36/ studyInstanceUID = '1.3.6.1.4.1.5962.99.1.2103930081.1286074986.1595536829665.3.0') shows differerent FrameOfReferenceUID between channels and also between metadata instances of the same channel. 

For the moment therefore multichannel are imposed if they have same geometry (Origins, Extents,Resolutions, GridSizes, TileSizes, PixelSpacings). 

Is there actually any advantage for checking also for the FrameOfReferenceUID? I mean if all the geometries checks are satisfied, we can actaully overlay the data.

### CODE REVIEW

**Comment by Punzo** [comment]
@hackermd please notice that I downgraded ol back to version 6.3.1. It looks that from version 6.4.0 (and also 6.5.0) they clear all the tiles cache for tiles not the view when zooming. This for us provides very bad perfomances. I didn't find an option in 6.5.0 to disable the tiles cache clearing.

**Comment by Punzo** [comment]
setting a  very large initial cacheSize value for the tile source seems to fix the tiles cache clearing in ol version 6.5.0

File: src/metadata.js
Comment (hackermd) [reply src/metadata.js:138]: See also https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-853365144

File: src/viewer.js
Comment (hackermd) [inline src/viewer.js:483]: This doesn't account for the option `{ useWebGL: false }`

File: src/viewer.js
Comment (hackermd) [inline src/viewer.js:492]: See comment on `useWegGL`

File: src/viewer.js
Comment (hackermd) [reply src/viewer.js:483]: Or in other words, the option may be `undefined` rather than `false` which is not ideal in my opinion.

File: src/viewer.js
Comment (hackermd) [inline src/viewer.js:542]: We cannot just group by Optical Path Identifier (see https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-853365144)

File: src/viewer.js
Comment (hackermd) [inline src/viewer.js:551]: Let's track such TODOs in issues rather than comments in the source code.

File: src/viewer.js
Comment (hackermd) [reply src/viewer.js:542]: Or we need to assert that each instance has a unique identifier. But what do we do if that's not the case? The best may be to introduce a Optical Path UID to the standard (see https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-853369764)

File: src/viewer.js
Comment (hackermd) [inline src/viewer.js:1134]: Not sure this is a good idea (see comment in `channel.js`).

**Review by hackermd — CHANGES_REQUESTED** [review]
@Punzo I already added a couple of comments and suggestions. However, before we finalize this PR, I would like to wait for @dclunie's feedback on https://github.com/MGHComputationalPathology/slim/issues/22

File: src/channel.js
Comment (Punzo) [reply src/channel.js:132]: yes I agree, now is commented otherwise it will not work with datasets server http://34.68.90.36/ studyInstanceUID = '1.3.6.1.4.1.5962.99.1.2103930081.1286074986.1595536829665.3.0 (and I can't show a demo of the new blending UI in slim). I will not apply/close this comment now, but I will apply it  as soon as we figure out the the grouping issue in https://github.com/MGHComputationalPathology/slim/issues/22.

File: src/channel.js
Comment (Punzo) [reply src/channel.js:471]: yes I agree, now is commented otherwise it will not work with datasets server http://34.68.90.36/ studyInstanceUID = '1.3.6.1.4.1.5962.99.1.2103930081.1286074986.1595536829665.3.0 (and I can't show a demo of the new blending UI in slim). I will not apply/close this comment now, but I will apply it as soon as we figure out the the grouping issue in MGHComputationalPathology/slim#22

File: src/channel.js
Comment (Punzo) [reply src/channel.js:672]: limitValues are interger, but yes it is always to check the numbers with a eps. Done

File: src/metadata.js
Comment (Punzo) [reply src/metadata.js:138]: yes I agree, I will not apply/close this comment now, but I will apply it as soon as we figure out the the grouping issue in MGHComputationalPathology/slim#22

File: src/channel.js
Comment (Punzo) [reply src/channel.js:412]: set to 100, restested and works well.

File: src/viewer.js
Comment (Punzo) [reply src/viewer.js:483]: no if at in the options of the constructor is passed, after the check it will be:
1) { useWebGL: false }, this[_options].useWebGL will be false
2) { useWebGL: true }, this[_options].useWebGL will be true
3) {  } (i.e. undefined), this[_options].useWebGL will be true (i.e. the default)

File: src/viewer.js
Comment (Punzo) [reply src/viewer.js:492]: see https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/54#discussion_r644539465

File: src/viewer.js
Comment (Punzo) [reply src/viewer.js:542]: yes I agree, I will not apply/close this comment now, but I will apply it as soon as we figure out the the grouping issue in MGHComputationalPathology/slim#22

File: src/viewer.js
Comment (Punzo) [reply src/viewer.js:551]: we already have it https://github.com/MGHComputationalPathology/slim/issues/11, I will just remove the TO DO line

File: src/viewer.js
Comment (Punzo) [reply src/viewer.js:1134]: set to 100, retested and works well.

**Comment by Punzo** [comment]
> @Punzo I already added a couple of comments and suggestions. However, before we finalize this PR, I would like to wait for @dclunie's feedback on [MGHComputationalPathology/slim#22](https://github.com/MGHComputationalPathology/slim/issues/22)

Thanks for the review! I have applied the minor comments in https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/54/commits/65821d5065654dbe7b4bd80496ea4bbb1f5b4419. I left open all the comments regarding the grouping of instances, Frame or reference, etc... we can fix them once we have David feedback.

**Comment by Punzo** [comment]
Added all the checks on frame of reference and container identifier, @hackermd please review. I tested locally (with my local DCM4CHEE) with the new data from David (i.e. https://github.com/MGHComputationalPathology/slim/issues/22#issuecomment-854919111). 

P.S.: at the moment the blend example will not work, because we have to update the Steve's open server (http://34.68.90.36/) with the new data.

**Comment by Punzo** [comment]
> P.S.: at the moment the blend example will not work, because we have to update the Steve's open server (http://34.68.90.36/) with the new data.

@hackermd upate, Steve updated the new data from David in http://34.68.90.36/, so everything ready to go.

File: src/channel.js
Comment (Punzo) [reply src/channel.js:241]: I don't think there would be issues, i.e. if you set needToRerender = true in antoher function, you would have simply that you will force the rendering and no harm is done. But in general I also don't see why other functions should set the needToRerender variable. The refresh of the rendering is done automatically when chaning visualization parameters, zooming or panning (everything by the updateTilesRendering function).

File: src/channel.js
Comment (Punzo) [reply src/channel.js:419]: done

File: src/viewer.js
Comment (Punzo) [reply src/viewer.js:598]: I don't think it would be difficult, but I need data to look at and to use for testing. I have a dded an issue tracker in slim https://github.com/MGHComputationalPathology/slim/issues/30

File: src/viewer.js
Comment (Punzo) [reply src/viewer.js:1171]: I think at the moment we can;t do anything, just hoping that in tge study the Optical Path Identifier is unique as it happened up to now with the data provided by David. I agree this is not a long-term solution. We should implement in the DICOM standard the OpticalPathUID as you previously proposed. I have added an issue tracker for this https://github.com/MGHComputationalPathology/slim/issues/31

File: src/viewer.js
Comment (Punzo) [reply src/viewer.js:611]: no because in teh case of only multichannels, this value can be zero and we don't have to fire any warning

**Comment by Punzo** [comment]
@hackermd thanks for the second review, I have replied/applied the comments!

**Review by hackermd — APPROVED** [review]

**Comment by hackermd** [comment]
Excellent! Thank you @Punzo

**Pull request merged**
